_posts/2019-07-26-pysam-with-s3.md

```diff
@@ -0,0 +1,194 @@
+---
+layout: blogpost
+title: Using PySAM to access private S3 Bucket
+author: Jackson Zheng
+date: 2019-07-26
+---
+
+I was working on a project for [CanDIG](https://www.distributedgenomics.ca/) which was an implementation of the [Htsget API](http://samtools.github.io/hts-specs/htsget.html) &mdash; a simple data-access API for reads and variants. A key component of our application was the [PySAM library](https://pysam.readthedocs.io/en/latest/), which provides an interface to genomic data. Its main use in the service was to parse a desired file and return only a chunk of that file with various filters. The most common way of accessing a file is with a local file path. However, with S3 buckets emerging as a popular file storage system, it is natural that PySAM should be able to interact with S3 objects; CanDIG, for instance, is moving to using a standard S3 interface for its data objects using [MinIO](https://min.io). PySAM is a wrapper of the htslib C-API which should support fetching of S3 objects, but many issues arose when my team and I tried to pass access objects in private S3 buckets with pysam or htslib. After many google searches, there seems to be many people having similar issues:
+1) https://www.biostars.org/p/213305/
+2) https://github.com/pysam-developers/pysam/issues/557
+3) https://www.biostars.org/p/213305/ 
+
+Although several sources claim that accessing s3 objects should just work, I have not been able to find a single worked example. After many days of digging on the web and coming across several problems, my team and I figured out how to successfully use PySAM and htslib to access private S3 bucket; we describe how below.
+
+### 1) Download and install htslib version from developer branch
+   
+We believe the root of the problem is AWS's move to a [new algorithm](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html) for signed URL access to private objects which is not supported by the current release of htslib (v1.9); but is supported on the `devel` branch.  To use this new version, the installation instructions can be found here: https://github.com/samtools/htslib . Make sure to add the --enable-s3 option during installation. 
+
+To check whether htslib installed correctly, use the command:
+```
+which htsfile
+```
+The output should be:
+```
+/usr/local/bin/htsfile
+```
+
+### 2) Try to Run PySAM
```

Comment:
Using PySAM here doesn't make sense, since it hasn't been rebuilt.  The point should be to show that with the new htslib, you can successfully access an AWS S3 bucket using just access keys and htsfile; using the example I sent over in slack, just demonstrate that you can set the environment variables below and access a private S3 bucket (like the htslibtest one I pointed out to you).  No minio at this point.